File: java/src/main/java/com/pqc/mldsa/MLDSA.java
package com.pqc.mldsa;

import com.pqc.common.Keccak;

import java.security.SecureRandom;
import java.util.Arrays;

import static com.pqc.mldsa.DsaField.*;
import static com.pqc.mldsa.DsaNTT.*;
import static com.pqc.mldsa.Decompose.*;
import static com.pqc.mldsa.DsaEncode.*;
import static com.pqc.mldsa.DsaHash.*;

/**
 * ML-DSA public API (FIPS 204).
 * Implements KeyGen, Sign, and Verify.
 */
public final class MLDSA {

    private MLDSA() {}

    public record KeyPair(byte[] pk, byte[] sk) {}

    // ========================================================================
    // KeyGen (FIPS 204 Algorithm 1)
    // ========================================================================

    public static KeyPair keyGen(DsaParams params) {
        SecureRandom rng = new SecureRandom();
        byte[] xi = new byte[32]; // random seed
        rng.nextBytes(xi);
        return keyGenInternal(xi, params);
    }

    /**
     * Deterministic key generation from seed (for testing).
     */
    public static KeyPair keyGenInternal(byte[] xi, DsaParams params) {
        int k = params.k;
        int l = params.l;

        // H(xi || k || l) -> rho, rhoPrime, K
        // FIPS 204: (rho, rho', K) = H(xi, 128) where H = SHAKE-256
        // Actually: expand xi with SHAKE-256 to get rho(32), rhoPrime(64), K(32)
        byte[] expanded = Keccak.shake256(xi, 128);
        byte[] rho = Arrays.copyOfRange(expanded, 0, 32);
        byte[] rhoPrime = Arrays.copyOfRange(expanded, 32, 96);
        byte[] K = Arrays.copyOfRange(expanded, 96, 128);

        // Generate matrix A (in NTT domain)
        int[][] aHat = expandA(rho, k, l);

        // Generate secret vectors s1, s2
        int[][] sAll = expandS(rhoPrime, k, l, params.eta);
        int[][] s1 = new int[l][];
        int[][] s2 = new int[k][];
        for (int i = 0; i < l; i++) s1[i] = sAll[i];
        for (int i = 0; i < k; i++) s2[i] = sAll[l + i];

        // NTT(s1)
        int[][] s1Hat = new int[l][];
        for (int i = 0; i < l; i++) {
            s1Hat[i] = ntt(s1[i]);
        }

        // t = A * s1 + s2
        int[][] t = new int[k][];
        for (int i = 0; i < k; i++) {
            int[] ti = new int[256]; // accumulator in NTT domain
            for (int j = 0; j < l; j++) {
                int[] product = pointwiseMul(aHat[i * l + j], s1Hat[j]);
                for (int c = 0; c < 256; c++) {
                    ti[c] = fieldAdd(ti[c], product[c]);
                }
            }
            ti = nttInverse(ti);
            // Add s2[i]
            t[i] = new int[256];
            for (int c = 0; c < 256; c++) {
                t[i][c] = fieldAdd(ti[c], s2[i][c]);
            }
        }

        // Power2Round: t -> (t1, t0)
        int[][] t1 = new int[k][256];
        int[][] t0 = new int[k][256];
        for (int i = 0; i < k; i++) {
            for (int j = 0; j < 256; j++) {
                int[] parts = power2Round(t[i][j]);
                t1[i][j] = parts[0];
                t0[i][j] = modQ(parts[1]);
            }
        }

        // Encode public key
        byte[] pk = encodePK(rho, t1, k);

        // tr = H(pk) = SHAKE-256(pk, 64)
        byte[] tr = Keccak.shake256(pk, 64);

        // Encode secret key
        byte[] sk = encodeSK(rho, K, tr, s1, s2, t0, params);

        return new KeyPair(pk, sk);
    }

    // ========================================================================
    // Sign (FIPS 204 Algorithm 2)
    // ========================================================================

    public static byte[] sign(byte[] sk, byte[] msg, DsaParams params) {
        // Decode secret key
        Object[] skParts = decodeSK(sk, params);
        byte[] rho = (byte[]) skParts[0];
        byte[] K = (byte[]) skParts[1];
        byte[] tr = (byte[]) skParts[2];
        int[][] s1 = (int[][]) skParts[3];
        int[][] s2 = (int[][]) skParts[4];
        int[][] t0 = (int[][]) skParts[5];

        int k = params.k;
        int l = params.l;

        // NTT of secret vectors
        int[][] s1Hat = new int[l][];
        for (int i = 0; i < l; i++) s1Hat[i] = ntt(s1[i]);
        int[][] s2Hat = new int[k][];
        for (int i = 0; i < k; i++) s2Hat[i] = ntt(s2[i]);
        int[][] t0Hat = new int[k][];
        for (int i = 0; i < k; i++) t0Hat[i] = ntt(t0[i]);

        // Expand A
        int[][] aHat = expandA(rho, k, l);

        // mu = H(tr || msg, 64) = SHAKE-256(tr || msg, 64)
        byte[] trMsg = concat(tr, msg);
        byte[] mu = Keccak.shake256(trMsg, 64);

        // rho'' = H(K || mu, 64) - for deterministic signing (no rnd)
        byte[] rhoPrimePrime = Keccak.shake256(concat(K, mu), 64);

        int kappa = 0;
        while (true) {
            // y = ExpandMask(rho'', kappa)
            int[][] y = expandMask(rhoPrimePrime, kappa, l, params.gamma1);

            // w = A * NTT(y)
            int[][] yHat = new int[l][];
            for (int i = 0; i < l; i++) yHat[i] = ntt(y[i]);

            int[][] w = new int[k][];
            for (int i = 0; i < k; i++) {
                int[] wi = new int[256];
                for (int j = 0; j < l; j++) {
                    int[] product = pointwiseMul(aHat[i * l + j], yHat[j]);
                    for (int c = 0; c < 256; c++) {
                        wi[c] = fieldAdd(wi[c], product[c]);
                    }
                }
                w[i] = nttInverse(wi);
            }

            // Decompose w into (w1, w0) where w1 = HighBits, w0 = LowBits
            int[][] w1 = new int[k][256];
            int[][] w0 = new int[k][256];
            for (int i = 0; i < k; i++) {
                for (int j = 0; j < 256; j++) {
                    int[] parts = decompose(w[i][j], params.gamma2);
                    w1[i][j] = parts[0];
                    w0[i][j] = parts[1];
                }
            }

            // c_tilde = H(mu || w1Encode, lambda/4)
            byte[] w1Encoded = encodeW1(w1, k, params.gamma2);
            byte[] hashInput = concat(mu, w1Encoded);
            int cTildeLen = params.lambda / 4;
            byte[] cTilde = Keccak.shake256(hashInput, cTildeLen);

            // c = SampleInBall(c_tilde)
            int[] cPoly = sampleInBall(cTilde, params.tau);
            int[] cHat = ntt(cPoly);

            // z = y + c * s1
            int[][] z = new int[l][];
            for (int i = 0; i < l; i++) {
                int[] cs1 = nttInverse(pointwiseMul(cHat, s1Hat[i]));
                z[i] = new int[256];
                for (int j = 0; j < 256; j++) {
                    z[i][j] = fieldAdd(y[i][j], cs1[j]);
                }
            }

            // Check z norm: ||z||_inf < gamma1 - beta
            if (polyVecNormExceeds(z, l, params.gamma1 - params.beta)) {
                kappa += l;
                continue;
            }

            // Compute cs2 and modify w0
            int[][] cs2 = new int[k][];
            for (int i = 0; i < k; i++) {
                cs2[i] = nttInverse(pointwiseMul(cHat, s2Hat[i]));
            }

            // w0 = w0 - cs2 (modify low bits)
            boolean reject = false;
            for (int i = 0; i < k; i++) {
                for (int j = 0; j < 256; j++) {
                    w0[i][j] = toSigned(fieldSub(w0[i][j] < 0 ? modQ(w0[i][j]) : w0[i][j], cs2[i][j]));
                }
            }

            // Check |w0 - cs2| < gamma2 - beta
            for (int i = 0; i < k; i++) {
                for (int j = 0; j < 256; j++) {
                    if (w0[i][j] >= params.gamma2 - params.beta || w0[i][j] <= -(params.gamma2 - params.beta)) {
                        reject = true;
                        break;
                    }
                }
                if (reject) break;
            }
            if (reject) {
                kappa += l;
                continue;
            }

            // Compute ct0
            int[][] ct0 = new int[k][];
            for (int i = 0; i < k; i++) {
                ct0[i] = nttInverse(pointwiseMul(cHat, t0Hat[i]));
            }

            // Check ct0 norm
            if (polyVecNormExceeds(ct0, k, params.gamma2)) {
                kappa += l;
                continue;
            }

            // w0 = w0 + ct0 (add ct0 to modified low bits)
            for (int i = 0; i < k; i++) {
                for (int j = 0; j < 256; j++) {
                    w0[i][j] = w0[i][j] + toSigned(ct0[i][j]);
                }
            }

            // Make hints: make_hint(w0, w1) per reference implementation
            int[][] hints = new int[k][256];
            int hintCount = 0;
            for (int i = 0; i < k; i++) {
                for (int j = 0; j < 256; j++) {
                    hints[i][j] = makeHint(w0[i][j], w1[i][j], params.gamma2);
                    hintCount += hints[i][j];
                }
            }

            if (hintCount > params.omega) {
                kappa += l;
                continue;
            }

            // Success - encode signature
            return encodeSig(cTilde, z, hints, params);
        }
    }

    // ========================================================================
    // Verify (FIPS 204 Algorithm 3)
    // ========================================================================

    public static boolean verify(byte[] pk, byte[] msg, byte[] sig, DsaParams params) {
        if (sig.length != params.sigSize) return false;
        if (pk.length != params.pkSize) return false;

        int k = params.k;
        int l = params.l;

        // Decode public key
        Object[] pkParts = decodePK(pk, k);
        byte[] rho = (byte[]) pkParts[0];
        int[][] t1 = (int[][]) pkParts[1];

        // Decode signature
        Object[] sigParts = decodeSig(sig, params);
        byte[] cTilde = (byte[]) sigParts[0];
        int[][] z = (int[][]) sigParts[1];
        int[][] h = (int[][]) sigParts[2];

        // Check z norm
        if (polyVecNormExceeds(z, l, params.gamma1 - params.beta)) {
            return false;
        }

        // Check hint count
        int hintCount = 0;
        for (int i = 0; i < k; i++) {
            for (int j = 0; j < 256; j++) {
                hintCount += h[i][j];
            }
        }
        if (hintCount > params.omega) return false;

        // Expand A
        int[][] aHat = expandA(rho, k, l);

        // tr = H(pk)
        byte[] tr = Keccak.shake256(pk, 64);
        byte[] mu = Keccak.shake256(concat(tr, msg), 64);

        // c = SampleInBall(cTilde)
        int[] cPoly = sampleInBall(cTilde, params.tau);
        int[] cHat = ntt(cPoly);

        // NTT(z)
        int[][] zHat = new int[l][];
        for (int i = 0; i < l; i++) zHat[i] = ntt(z[i]);

        // w'_approx = A*z - c*t1*2^d (all in NTT domain)
        // t1 * 2^d in NTT: ntt(t1[i] << 13)
        int[][] w1Prime = new int[k][256];
        for (int i = 0; i < k; i++) {
            // A[i]*z
            int[] azHat = new int[256];
            for (int j = 0; j < l; j++) {
                int[] product = pointwiseMul(aHat[i * l + j], zHat[j]);
                for (int c = 0; c < 256; c++) {
                    azHat[c] = fieldAdd(azHat[c], product[c]);
                }
            }

            // c * t1[i] * 2^d in NTT
            int[] t1Scaled = new int[256];
            for (int j = 0; j < 256; j++) {
                t1Scaled[j] = fieldMul(t1[i][j], 1 << 13);
            }
            int[] t1ScaledHat = ntt(t1Scaled);
            int[] ct1Hat = pointwiseMul(cHat, t1ScaledHat);

            // w_approx = A*z - c*t1*2^d
            int[] wApprox = new int[256];
            for (int c = 0; c < 256; c++) {
                wApprox[c] = fieldSub(azHat[c], ct1Hat[c]);
            }
            int[] wApproxTime = nttInverse(wApprox);

            // UseHint to recover w1
            for (int j = 0; j < 256; j++) {
                w1Prime[i][j] = useHint(h[i][j], wApproxTime[j], params.gamma2);
            }
        }

        // Recompute c_tilde
        byte[] w1Encoded = encodeW1(w1Prime, k, params.gamma2);
        byte[] hashInput = concat(mu, w1Encoded);
        int cTildeLen = params.lambda / 4;
        byte[] cTildeCheck = Keccak.shake256(hashInput, cTildeLen);

        return Arrays.equals(cTilde, cTildeCheck);
    }

    // ========================================================================
    // Utility functions
    // ========================================================================

    /**
     * Check if any coefficient in the polynomial vector exceeds the bound.
     */
    private static boolean polyVecNormExceeds(int[][] v, int len, int bound) {
        for (int i = 0; i < len; i++) {
            for (int j = 0; j < 256; j++) {
                int val = toSigned(v[i][j]);
                if (val <= -bound || val >= bound) {
                    return true;
                }
            }
        }
        return false;
    }

    private static byte[] concat(byte[] a, byte[] b) {
        byte[] result = new byte[a.length + b.length];
        System.arraycopy(a, 0, result, 0, a.length);
        System.arraycopy(b, 0, result, a.length, b.length);
        return result;
    }
}

File: dotnet/src/PqcStandards/MlDsa/MlDsa.cs
using System.Security.Cryptography;
using PqcStandards.Common;

namespace PqcStandards.MlDsa;

/// <summary>ML-DSA digital signature algorithm (FIPS 204).</summary>
public static class MlDsaAlgorithm
{
    /// <summary>ML-DSA.KeyGen: returns (pk, sk).</summary>
    public static (byte[] pk, byte[] sk) KeyGen(MlDsaParams p)
    {
        byte[] xi = RandomNumberGenerator.GetBytes(32);
        return KeyGenInternal(p, xi);
    }

    internal static (byte[] pk, byte[] sk) KeyGenInternal(MlDsaParams p, byte[] xi)
    {
        // (rho, rhoPrime, K) = H(xi || k || l)
        byte[] input = new byte[xi.Length + 2];
        Buffer.BlockCopy(xi, 0, input, 0, xi.Length);
        input[xi.Length] = (byte)p.K;
        input[xi.Length + 1] = (byte)p.L;
        byte[] h = Keccak.Shake256(input, 128);
        byte[] rho = h[..32];
        byte[] rhoPrime = h[32..96];
        byte[] K = h[96..128];

        // A_hat = ExpandA(rho)
        int[][] Ahat = DsaHash.ExpandA(rho, p.K, p.L);

        // (s1, s2) = ExpandS(rhoPrime)
        var (s1, s2) = DsaHash.ExpandS(rhoPrime, p.K, p.L, p.Eta);

        // s1_hat = NTT(s1)
        int[][] s1hat = new int[p.L][];
        for (int i = 0; i < p.L; i++)
            s1hat[i] = DsaNtt.NttForward(s1[i]);

        // t = NTT^-1(A_hat * s1_hat) + s2
        int[][] t = new int[p.K][];
        for (int i = 0; i < p.K; i++)
        {
            t[i] = new int[256];
            for (int j = 0; j < p.L; j++)
            {
                int[] prod = DsaNtt.PointwiseMul(Ahat[i * p.L + j], s1hat[j]);
                for (int c = 0; c < 256; c++)
                    t[i][c] = DsaField.Add(t[i][c], prod[c]);
            }
            t[i] = DsaNtt.NttInverse(t[i]);
            for (int c = 0; c < 256; c++)
                t[i][c] = DsaField.Add(t[i][c], s2[i][c]);
        }

        // (t1, t0) = Power2Round(t)
        int[][] t1 = new int[p.K][];
        int[][] t0 = new int[p.K][];
        for (int i = 0; i < p.K; i++)
        {
            t1[i] = new int[256];
            t0[i] = new int[256];
            for (int j = 0; j < 256; j++)
            {
                var (hi, lo) = Decompose.Power2Round(t[i][j]);
                t1[i][j] = hi;
                t0[i][j] = DsaField.ModQ(lo);
            }
        }

        byte[] pk = DsaEncode.EncodePk(rho, t1, p.K);
        byte[] tr = Keccak.Shake256(pk, 64);
        byte[] sk = DsaEncode.EncodeSk(rho, K, tr, s1, s2, t0, p.K, p.L, p.Eta);

        return (pk, sk);
    }

    /// <summary>ML-DSA.Sign: returns signature sigma.</summary>
    public static byte[] Sign(MlDsaParams p, byte[] sk, byte[] message)
    {
        var (rho, K, tr, s1, s2, t0) = DsaEncode.DecodeSk(sk, p.K, p.L, p.Eta);

        // mu = H(tr || M)
        byte[] muInput = new byte[tr.Length + message.Length];
        Buffer.BlockCopy(tr, 0, muInput, 0, tr.Length);
        Buffer.BlockCopy(message, 0, muInput, tr.Length, message.Length);
        byte[] mu = Keccak.Shake256(muInput, 64);

        // rho' = H(K || mu)  (deterministic signing)
        byte[] rhoP = new byte[K.Length + mu.Length];
        Buffer.BlockCopy(K, 0, rhoP, 0, K.Length);
        Buffer.BlockCopy(mu, 0, rhoP, K.Length, mu.Length);
        byte[] rhoPrime = Keccak.Shake256(rhoP, 64);

        // A_hat = ExpandA(rho)
        int[][] Ahat = DsaHash.ExpandA(rho, p.K, p.L);

        // NTT of s1, s2, t0
        int[][] s1hat = new int[p.L][];
        int[][] s2hat = new int[p.K][];
        int[][] t0hat = new int[p.K][];
        for (int i = 0; i < p.L; i++)
            s1hat[i] = DsaNtt.NttForward(s1[i]);
        for (int i = 0; i < p.K; i++)
        {
            s2hat[i] = DsaNtt.NttForward(s2[i]);
            t0hat[i] = DsaNtt.NttForward(t0[i]);
        }

        // Rejection sampling loop
        int kappa = 0;
        while (true)
        {
            // y = ExpandMask(rhoPrime, kappa)
            int[][] y = DsaHash.ExpandMask(rhoPrime, kappa, p.L, p.Gamma1);
            kappa += p.L;

            // w = NTT^-1(A_hat * NTT(y))
            int[][] yhat = new int[p.L][];
            for (int i = 0; i < p.L; i++)
                yhat[i] = DsaNtt.NttForward(y[i]);

            int[][] w = new int[p.K][];
            for (int i = 0; i < p.K; i++)
            {
                w[i] = new int[256];
                for (int j = 0; j < p.L; j++)
                {
                    int[] prod = DsaNtt.PointwiseMul(Ahat[i * p.L + j], yhat[j]);
                    for (int ci = 0; ci < 256; ci++)
                        w[i][ci] = DsaField.Add(w[i][ci], prod[ci]);
                }
                w[i] = DsaNtt.NttInverse(w[i]);
            }

            // w1 = HighBits(w)
            int[][] w1 = new int[p.K][];
            for (int i = 0; i < p.K; i++)
            {
                w1[i] = new int[256];
                for (int j = 0; j < 256; j++)
                    w1[i][j] = Decompose.HighBits(w[i][j], p.Gamma2);
            }

            // c_tilde = H(mu || EncodeW1(w1))
            byte[] w1Enc = DsaEncode.EncodeW1(w1, p.K, p.Gamma2);
            byte[] cInput = new byte[mu.Length + w1Enc.Length];
            Buffer.BlockCopy(mu, 0, cInput, 0, mu.Length);
            Buffer.BlockCopy(w1Enc, 0, cInput, mu.Length, w1Enc.Length);
            byte[] cTilde = Keccak.Shake256(cInput, p.Lambda / 4);

            // c = SampleInBall(c_tilde)
            int[] c = DsaHash.SampleInBall(cTilde, p.Tau);
            int[] chat = DsaNtt.NttForward(c);

            // z = y + NTT^-1(c_hat * s1_hat)
            int[][] z = new int[p.L][];
            for (int i = 0; i < p.L; i++)
            {
                int[] cs1 = DsaNtt.NttInverse(DsaNtt.PointwiseMul(chat, s1hat[i]));
                z[i] = new int[256];
                for (int j = 0; j < 256; j++)
                    z[i][j] = DsaField.Add(y[i][j], cs1[j]);
            }

            // Check ||z||_inf < gamma1 - beta
            if (!CheckNormBound(z, p.L, p.Gamma1 - p.Beta))
                continue;

            // r0 = LowBits(w - NTT^-1(c_hat * s2_hat))
            bool reject = false;
            int[][] h = new int[p.K][];
            int hintCount = 0;
            for (int i = 0; i < p.K; i++)
            {
                int[] cs2 = DsaNtt.NttInverse(DsaNtt.PointwiseMul(chat, s2hat[i]));
                int[] r0 = new int[256];
                for (int j = 0; j < 256; j++)
                {
                    int wMinusCss2 = DsaField.Sub(w[i][j], cs2[j]);
                    r0[j] = Decompose.LowBits(wMinusCss2, p.Gamma2);
                }
                // Check ||r0||_inf < gamma2 - beta
                for (int j = 0; j < 256; j++)
                {
                    int centered = DsaField.CenterMod(DsaField.ModQ(r0[j]));
                    if (Math.Abs(centered) >= p.Gamma2 - p.Beta)
                    { reject = true; break; }
                }
                if (reject) break;

                // Compute hint
                int[] ct0 = DsaNtt.NttInverse(DsaNtt.PointwiseMul(chat, t0hat[i]));
                h[i] = new int[256];
                for (int j = 0; j < 256; j++)
                {
                    int wPrime = DsaField.Sub(w[i][j], cs2[j]);
                    h[i][j] = Decompose.MakeHint(DsaField.ModQ(ct0[j]), wPrime, p.Gamma2);
                    hintCount += h[i][j];
                }
            }

            if (reject || hintCount > p.Omega)
                continue;

            // Check ct0 norm
            bool ct0Reject = false;
            for (int i = 0; i < p.K && !ct0Reject; i++)
            {
                int[] ct0 = DsaNtt.NttInverse(DsaNtt.PointwiseMul(chat, t0hat[i]));
                for (int j = 0; j < 256; j++)
                {
                    int centered = DsaField.CenterMod(ct0[j]);
                    if (Math.Abs(centered) >= p.Gamma2)
                    { ct0Reject = true; break; }
                }
            }
            if (ct0Reject) continue;

            return DsaEncode.EncodeSig(cTilde, z, h, p);
        }
    }

    /// <summary>ML-DSA.Verify: returns true if signature is valid.</summary>
    public static bool Verify(MlDsaParams p, byte[] pk, byte[] message, byte[] sig)
    {
        if (sig.Length != p.SigSize) return false;

        var (rho, t1) = DsaEncode.DecodePk(pk, p.K);
        byte[] tr = Keccak.Shake256(pk, 64);

        // mu = H(tr || M)
        byte[] muInput = new byte[tr.Length + message.Length];
        Buffer.BlockCopy(tr, 0, muInput, 0, tr.Length);
        Buffer.BlockCopy(message, 0, muInput, tr.Length, message.Length);
        byte[] mu = Keccak.Shake256(muInput, 64);

        var decoded = DsaEncode.DecodeSig(sig, p);
        if (decoded == null) return false;
        var (cTilde, z, h) = decoded.Value;

        // Check z norm
        if (!CheckNormBound(z, p.L, p.Gamma1 - p.Beta))
            return false;

        // A_hat = ExpandA(rho)
        int[][] Ahat = DsaHash.ExpandA(rho, p.K, p.L);

        // c from c_tilde
        int[] c = DsaHash.SampleInBall(cTilde, p.Tau);
        int[] chat = DsaNtt.NttForward(c);

        // NTT(z)
        int[][] zhat = new int[p.L][];
        for (int i = 0; i < p.L; i++)
            zhat[i] = DsaNtt.NttForward(z[i]);

        // w'_approx = NTT^-1(A_hat * NTT(z) - c_hat * NTT(t1 * 2^d))
        int[][] w1Prime = new int[p.K][];
        for (int i = 0; i < p.K; i++)
        {
            int[] az = new int[256];
            for (int j = 0; j < p.L; j++)
            {
                int[] prod = DsaNtt.PointwiseMul(Ahat[i * p.L + j], zhat[j]);
                for (int c2 = 0; c2 < 256; c2++)
                    az[c2] = DsaField.Add(az[c2], prod[c2]);
            }

            // t1 * 2^13
            int[] t1Shifted = new int[256];
            for (int j = 0; j < 256; j++)
                t1Shifted[j] = DsaField.Mul(t1[i][j], 1 << 13);
            int[] t1hat = DsaNtt.NttForward(t1Shifted);
            int[] ct1 = DsaNtt.PointwiseMul(chat, t1hat);

            int[] wApprox = new int[256];
            for (int j = 0; j < 256; j++)
                wApprox[j] = DsaField.Sub(az[j], ct1[j]);
            wApprox = DsaNtt.NttInverse(wApprox);

            // UseHint to recover w1'
            w1Prime[i] = new int[256];
            for (int j = 0; j < 256; j++)
                w1Prime[i][j] = Decompose.UseHint(h[i][j], wApprox[j], p.Gamma2);
        }

        // c_tilde' = H(mu || EncodeW1(w1'))
        byte[] w1Enc = DsaEncode.EncodeW1(w1Prime, p.K, p.Gamma2);
        byte[] cInput = new byte[mu.Length + w1Enc.Length];
        Buffer.BlockCopy(mu, 0, cInput, 0, mu.Length);
        Buffer.BlockCopy(w1Enc, 0, cInput, mu.Length, w1Enc.Length);
        byte[] cTildePrime = Keccak.Shake256(cInput, p.Lambda / 4);

        return CryptographicOperations.FixedTimeEquals(cTilde, cTildePrime);
    }

    private static bool CheckNormBound(int[][] vecs, int count, int bound)
    {
        for (int i = 0; i < count; i++)
            for (int j = 0; j < 256; j++)
            {
                int centered = DsaField.CenterMod(vecs[i][j]);
                if (Math.Abs(centered) >= bound)
                    return false;
            }
        return true;
    }
}
